### PR TITLE
(WIP) goTo() function to smooth scroll to the given location

### PR DESCRIPTION
**Pull request by zaken**
## Description
Easily smooth scroll to a given page offset / selector / DOM element.

## Markup:
`<v-btn @click="$vuetify.goTo(123, <offset>, <duration>)">scroll</v-btn>`
`<v-btn @click="$vuetify.goTo('#my-id', <offset>, <duration>)">scroll</v-btn>`
`<v-btn @click="$vuetify.goTo('.my-class', <offset>, <duration>)">scroll</v-btn>`
`<v-btn @click="$vuetify.goTo($refs.myRef, <offset>, <duration>)">scroll</v-btn>`

- Target must be a Number, Selector or a DOM Element.
- Offset is optional and must be a Number. Defaults to 0.
- Duration is optional and must be a Number. Defaults to 500.

## Playground:
https://hastebin.com/levamufusi.xml

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have created a PR in the [documentation](https://github.com/vuetifyjs/vuetifyjs.com) with the necessary changes.

### CODE REVIEW

**Comment by zaken** [comment]
Haven't dealt with typescript before. Type definition is probably completely wrong.

File: index.d.ts [hunk 1]
```diff
@@ -45,12 +45,16 @@ declare class VuetifyTheme {
...
+declare function goTo(n: number): void;
+declare function goTo(s: string): void;
+
 declare class VuetifyObject {
   application: VuetifyApplication;
   breakpoint: VuetifyBreakpoint;
   dark: boolean;
   theme: VuetifyTheme;
   touchSupport: boolean;
+  goTo: goTo;
```
Comment (KaelWD) [inline index.d.ts]: `goTo: (target: string | number | HTMLElement) => void;`

**Comment by jacekkarczmarczyk** [comment]
Is it good idea to include it in the core? I'd rather see it as a optional feature which - in case someone wants to use id - could be imported separately similar to `colors.js`

**Comment by johnleider** [comment]
Excellent PR. My opinion sides with @jacekkarczmarczyk on not including it with the core. This will lead to component functionality being directly tied to it instead of simply offering the ability to smooth things up. I'd be interested in seeing if we could come up with a way for both worlds.

Optional import but components can be hooked into it if included.

**Comment by nekosaur** [comment]
Use the vuetify console methods if you need to output any warnings or errors. Found in `util/console`

**Comment by johnleider** [comment]
Everything looks good, it just needs to be made as an optional import now.

**Comment by zaken** [comment]
Should it be a mixin?

File: index.d.ts [hunk 2]
```diff
@@ -60,12 +60,16 @@ declare interface VuetifyOptions {
   themeCache?: VuetifyThemeCache
 }
 
+declare function goTo(n: number): void;
+declare function goTo(s: string): void;
```
Comment (KaelWD) [inline index.d.ts]: Remove these

**Comment by KaelWD** [comment]
No, just a function. Put it in /util

File: src/components/Vuetify/mixins/goTo.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * Modified from https://github.com/alamcordeiro/vue-smooth-scroll
+ */
+
+function easeInOutCubic (t) { return t < .5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1 }
+
+export default function goTo (target, offset = 0) {
+  if (typeof window === 'undefined') return
+
+  if (offset && isNaN(offset)) {
+    consoleError(`Offset must be a number, received '${typeof offset}' instead.`)
```
Comment (KaelWD) [inline src/components/Vuetify/mixins/goTo.js]: Use `offset.constructor.name` instead of `typeof`

File: src/components/Vuetify/mixins/goTo.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  let end
+
+  if (typeof target === 'string') {
+    let anchor = target.charAt(0) === '#' ? target.substring(1) : target
+    let scrollTo = document.getElementById(anchor)
+    if(!scrollTo) return
+    end = scrollTo.getBoundingClientRect().top + window.pageYOffset
+  }
+  else if (typeof target === 'number') end = target
```
Comment (KaelWD) [inline src/components/Vuetify/mixins/goTo.js]: Add braces

File: src/components/Vuetify/mixins/goTo.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  end += offset
+  let duration = 500
+  let start = performance.now()
+
+  function step (now) {
+    let elapsed = now - start
+    let position = end
+    if (elapsed < duration){
```
Comment (KaelWD) [inline src/components/Vuetify/mixins/goTo.js]: Add a space after the closing bracket

File: src/util/goTo.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    consoleError(`Offset must be a Number, received ${offset.constructor.name} instead.`)
+    return
+  }
+
+  let end
+
+  if (typeof target === 'string') {
+    let anchor = target.charAt(0) === '#' ? target.substring(1) : target
+    let scrollTo = document.getElementById(anchor)
+    if(!scrollTo) return
```
Comment (jacekkarczmarczyk) [inline src/util/goTo.js]: space after `if`, `yarn lint` should want you about it, if it didn't then it's something that should be fixed

File: src/util/goTo.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  else if (typeof target === 'number') {
+    end = target
+  }
+  else {
+    consoleError(`Target must be a String/Number, received ${target.constructor.name} instead.`)
+    return
+  }
+
+  end += offset
+  let duration = 500
```
Comment (jacekkarczmarczyk) [inline src/util/goTo.js]: Don't like the hardcoded magic value, this should be a function argument

File: src/util/goTo.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  if (offset && isNaN(offset)) {
+    consoleError(`Offset must be a Number, received ${offset.constructor.name} instead.`)
+    return
+  }
+
+  let end
+
+  if (typeof target === 'string') {
+    let anchor = target.charAt(0) === '#' ? target.substring(1) : target
+    let scrollTo = document.getElementById(anchor)
```
Comment (jacekkarczmarczyk) [inline src/util/goTo.js]: Maybe just `document.querySelector(target)`?

File: src/util/goTo.js (hunk 6 above)
Comment (KaelWD) [reply src/util/goTo.js]: There's a few things in this PR that should be linting errors, but travis hasn't picked any of them up for some reason.

File: src/util/goTo.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    if(!scrollTo) return
+    end = scrollTo.getBoundingClientRect().top + window.pageYOffset
+  }
+  else if (typeof target === 'number') {
+    end = target
+  }
+  else {
```
Comment (KaelWD) [inline src/util/goTo.js]: Use 1tbs, stroustrup is gross

Comment (zaken) [reply src/util/goTo.js]: Yeah, somehow missed this..

**Comment by zaken** [comment]
Should we log an error if queried element not found?

**Comment by jacekkarczmarczyk** [comment]
You could remove the `if (!scrollTo) return` so that the next line would trigger easily trackable exception

And btw - target could be also a dom element, this might be quite a common case, for example when you want to scroll to some element from your component (`$refs.something.$el.children[3]`) which probably won't have any id or simple selector

**Pull request merged**